### PR TITLE
Ocr mobile

### PR DESCRIPTION
**Pull request by EdvardasDlugauskas**
Many renames and reorderings in CameraActivity.cs, some functionality separated to CameraActivity.Helpers.cs (made the class partial).

Actual changes mostly in RecogniseImage method.

### CODE REVIEW

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs [hunk 1]
```diff
@@ -187,127 +176,113 @@ protected override void OnActivityResult(int requestCode, Result resultCode, Int
...
+                    var ocrResult = recognitionResult.RawOcrResult;
+                    if (!string.IsNullOrEmpty(ocrResult))
+                        new MessageBarCreator(rootView, $"OCR: {ocrResult}").Show();
+
+                    GetShoppingCartAddItemDialog(recognitionResult).Show();
                 }, TaskScheduler.FromCurrentSynchronizationContext());
         }
 
-        public void AddToShoppingCart()
+        private void AddToShoppingCart(string guess)
```
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.cs:248]: @s1712333 I rewrote this method a bit, have a look

File: Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        private bool IsThereAnAppToTakePictures()
+        {
+            var intent = new Intent(MediaStore.ActionImageCapture);
+            var availableActivities =
+                PackageManager.QueryIntentActivities(intent, PackageInfoFlags.MatchDefaultOnly);
+            return availableActivities != null && availableActivities.Count > 0;
+        }
+
+        private ErrorDialogCreator GetShoppingCartAddItemDialog(RecognitionResult recognitionResult)
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs:35]: For some reason VS doesn't recognize the type RecognitionResult for me, I tried clean/build and no luck.

Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs:35]: Fixed

File: Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+using Android.Speech;
+using Camera;
+using ImageRecognitionMobile;
+using ImageRecognitionMobile.Classificators;
+using PCLAppConfig;
+using ShopLens.Droid.Helpers;
+using ShopLens.Droid.Notifications;
+using Unity;
+
+// ReSharper disable once CheckNamespace
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs]: Did Resharper automagically create this class for you? That's cool. 
Is this comment used by Resharper in any way?

File: Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
+﻿using System.Collections.Generic;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs:1]: You can create Helper classes like this? 
So if you want to create a helper for a specific activity you would create a class SpecificActivity.Helpers and that's it?
What are the advantages of doing this over just creating a static class in our project's Helpers folder?

File: Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            var task = Task.Run(() =>
+            {
+                var frame = new Frame.Builder().SetBitmap(image).Build();
+                var items = textRecognizer.Detect(frame);
+                var ocrResultBuilder = new StringBuilder();
+                for (var i = 0; i < items.Size(); i++)
+                {
+                    var item = (TextBlock) items.ValueAt(i);
+                    ocrResultBuilder.Append(item.Value);
+                    ocrResultBuilder.Append("/");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs]: I think "|" would look better in the snackbar, but that's just my opinion.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs (hunk 4 above)
Comment (marjasi) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs:1]: Oh okay, so you're not creating a Helper per say, you're just splitting the class's code into two files.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs [hunk 6]
```diff
@@ -19,51 +17,50 @@
...
+
+        private File productPhoto;
+        private File dir =
+            new File(Android.OS.Environment.GetExternalStoragePublicDirectory(Android.OS.Environment.DirectoryPictures),
+                "DCIM");
+
+        private static readonly string FileProviderName = ConfigurationManager.AppSettings["fileProviderName"];
+        private readonly string prefsName = ConfigurationManager.AppSettings["ShopCartPrefs"];
+        
+        private static readonly int RequestImageId = (int) ActivityIds.ImageRequest;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.cs:55]: I need to remember to start using spaces after casts to keep our code consistent.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs [hunk 7]
```diff
@@ -187,127 +176,113 @@ protected override void OnActivityResult(int requestCode, Result resultCode, Int
...
+                    {
+                        System.Diagnostics.Debug.WriteLine("[Warning]: Text recognizer not operational.");
+                        return new RecognitionResult { Predictions = await classifyImageTask };
+                    }
                     
-                }
-            })
+                    var recognizeTextTask = GetTextFromImageAsync(image);
+
+                    await Task.WhenAll(classifyImageTask, recognizeTextTask);
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.cs:208]: Why is this line needed?

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs [hunk 8]
```diff
@@ -187,127 +176,113 @@ protected override void OnActivityResult(int requestCode, Result resultCode, Int
...
                 }, TaskScheduler.FromCurrentSynchronizationContext());
         }
 
-        public void AddToShoppingCart()
+        private void AddToShoppingCart(string guess)
         {
-            prefs.AddString(guess.First().ToString().ToUpper() + guess.Substring(1));
-            shoppingCartMessageBar.Show();
+            var preferences = new ActivityPreferences(this, prefsName);
+            preferences.AddString(guess.FirstCharToUpper());
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.cs:251]: What? We actually needed that extension method somewhere else other that Shopping Cart?! 😃 nice!

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs [hunk 9]
```diff
@@ -187,127 +176,113 @@ protected override void OnActivityResult(int requestCode, Result resultCode, Int
...
+            shoppingCartItemAddedMessageBar.Show();
         }
 
         // When the current voice recognition session stops.
         public void OnResults(Bundle results)
         {
             var matches = results.GetStringArrayList(SpeechRecognizer.ResultsRecognition);
-            if (matches.Count > 0)
+            if (matches.Count <= 0) return;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.cs:259]: I always write the brackets with the if statement, even when its one line. I guess this writing style is better.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs [hunk 10]
```diff
@@ -19,51 +17,50 @@
...
+        private static readonly string FileProviderName = ConfigurationManager.AppSettings["fileProviderName"];
+        private readonly string prefsName = ConfigurationManager.AppSettings["ShopCartPrefs"];
+        
+        private static readonly int RequestImageId = (int) ActivityIds.ImageRequest;
+        private static readonly int RequestPermissionId = (int) ActivityIds.PermissionRequest;
+        private static readonly int PickImageId = (int) ActivityIds.PickImageRequest;
+        private static readonly string CmdWhatIsThis = ConfigurationManager.AppSettings["CmdWhatIsThis"];
+        private static readonly string CmdPickPhoto = ConfigurationManager.AppSettings["CmdPickPhoto"];
 
-        private static readonly int REQUEST_IMAGE = (int)ActivityIds.ImageRequest;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/CameraActivity.cs:62]: That's some Java like constant naming right there.

Comment (marjasi) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.cs:62]: I probably wrote those names.

File: Mobile/ShopLens/ImageRecognitionMobile/RecognitionResult.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace ImageRecognitionMobile
+{
+    public class RecognitionResult
+    {
+        public Dictionary<string, float> Predictions { get; set; } = new Dictionary<string, float>();
+        public string WeightSpecifier { get; set; } = "";
```
Comment (marjasi) [inline Mobile/ShopLens/ImageRecognitionMobile/RecognitionResult.cs:9]: You can set an initial value for a get set field?! 😮

File: Mobile/ShopLens/ImageRecognitionMobile/RecognitionResult.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        public string RawOcrResult { get; set; } = "";
```
Comment (marjasi) [inline Mobile/ShopLens/ImageRecognitionMobile/RecognitionResult.cs:10]: I also started using "" for empty strings today.

**Review by marjasi — APPROVED** [review]
Good job 👍. I left some comments but they're just about me asking stuff rather than pointing out any mistakes, so I approve. I would like to get answers to some, though.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs (hunk 3 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.Helpers.cs]: Yeah, ReSharper does use the comment. Perhaps it should be removed though

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs (hunk 6 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.cs:55]: Should probably create a standard and have it as a shared file.

Can alternatively use the .editorconfig file.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs (hunk 7 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.cs:208]: We have 2 running tasks at this point which are not awaited

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs (hunk 8 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.cs:251]: Yeah, Tomas didn't notice we have the method for that.

File: Mobile/ShopLens/Droid/Activities/CameraActivity.cs (hunk 9 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Activities/CameraActivity.cs:259]: Some reviewers always say it's better to write braces, others say it's better to omit braces. Probably should be noted in the coding standard.

File: Mobile/ShopLens/ImageRecognitionMobile/RecognitionResult.cs (hunk 12 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/ImageRecognitionMobile/RecognitionResult.cs:10]: Some say `string.Empty()` is more readable
